### ISSUE TITLE
Generalize AvalancheDataset __getitem__ for Handling General Y Tensors

### ISSUE THREAD

**Issue by vlomonaco** [ContinualAI/avalanche#372]
We should better generalize the `AvalancheDataset` __getitem__ so that it can handle any task (regression, segmentation, detection, ...) with arbitrary y tensors (not just integer).

**Comment by AntonioCarta** [comment]
I think pytorch dataloaders have a nice solution to this problem. They allow you to define a custom `collate_fn` that takes the single samples (in a list) and combines them, returning a mini-batch. We can have a default behavior (collate each element of the tuple by stacking along dim 0) and the provide a way for the user to customize the `collate_fn`.
